### PR TITLE
convert API to use json strings instead of json-c specific json_objects

### PR DESCRIPTION
**Pull request by garlick**
This is a tentative start at converting the Flux API to use JSON strings rather than JSON objects (see issue #169).  I'd like to get a bit of feedback if possible before I go further.  In this PR:

* The low level message accessors are updated to operate (only) on JSON strings (unit tests added)
* The higher level event functions in `event.h` were reworked to operate on JSON strings, and also to have distinct message encode/decode functions and message send/recv functions rather than blending the two.  (unit tests added)
* The wire encoding of JSON payloads was changed to include a C terminating NULL so that accessors could return pointers into the `zmsg_t` rather than having to duplicate payloads on the critical path, and requirig the user to free returned strings.
* Same applied to the topic string accessor and the representation of topic strings on the wire.
* Updated all users of these functions in flux-core

(Possibly once we have our own `flux_msg_t` we can drop the extra null from the wire again, as that feels a bit wrong; however the API improvement that this enables right now is IMHO very good)

The new message accessors look like this:
```
/* Get/set json string payload.
 * set allows s to be NULL
 * get will set *s to NULL and return success if there is no payload.
 */
int flux_msg_set_payload_json_str (zmsg_t *zmsg, const char *s);
int flux_msg_get_payload_json_str (zmsg_t *zmsg, const char **s);

int flux_msg_get_topic (zmsg_t *zmsg, const char **topic); // topic is now a const
```
And the event functions look like this:
```
 
/* Decode an event message.
 * If topic is non-NULL, assign the event topic string.
 * If json_str is non-NULL, assign the payload.  json_str indicates whether
 * payload is expected and it is an EPROTO error if they don't match.
 * Returns 0 on success, or -1 on failure with errno set.
 */
int flux_event_decode (zmsg_t *zmsg, const char **topic, const char **json_str);

/* Encode an event message.
 * If json_str is non-NULL, copy it to the message payload.
 * In the varargs variants, json_str is constructed from the arguments.
 * Returns message or NULL on failure with errno set.
 */
zmsg_t *flux_event_encode (const char *topic, const char *json_str);
zmsg_t *flux_event_vencodef (const char *topic, const char *fmt, va_list ap);
zmsg_t *flux_event_encodef (const char *topic, const char *fmt, ...);

/* Send/receive events
 */
int flux_event_send (flux_t h, zmsg_t **zmsg);
zmsg_t *flux_event_recv (flux_t h, bool nonblock);
```

If this direction seems good, I could continue on into the more complicated request/response/rpc functions but I thought I should pause here and iterate if I haven't got it quite right.

### CODE REVIEW

**Comment by grondo** [comment]
I don't really have any comments as this feels pretty straightforward. 

I do feel like the function naming `flux_msg_set_payload_json_str` is a bit clunky (there is no equivalent for `flux_event_*`, using `str` instead of `string`...). This is just a naming preference thing, but it feels like trailing `_str` at least is unneeded since it is clear from parameters this is a json string. (Also, this would mirror the calls for events)

Also, I kind of wonder if there is any benefit to making this change before introducing `flux_msg_t` -- does this pave the way for that work?

You converted a lot of API users... did it seem like this introduced a lot of verbosity? I'm wondering if introducing an in-tree convenience library that provides the old style calls (directly get json_object from flux messages) would cut down on a little bit of duplication?

**Comment by garlick** [comment]
Agreed we should drop the `_str` from `flux_msg_set_payload_json_str()` and `flux_msg_get_payload_json_str()` now that (in 2cb3be0) we've got rid of the existing functions with that name.

I'm not sure what you mean about there being no equivalent `flux_event_*` using str instead of string?

This does pave the way for `flux_msg_t`.  We could avoid some churn by putting them both in the same PR, but that would be a huge PR and I'd prefer to keep them separate.  It _would_ reduce the churn to create an in-tree convenience library as you suggest.  I think I should rework this PR with a `compat.[ch]` shim, however I would argue that it should be removed and users updated after the API has settled, since if the public API is annoyingly verbose, maybe we haven't got it right, and if flux-core programs to wrappers, we won't be providing clean examples of how to use the API.

I personally feel that the verbosity introduced by having a clear division of labor between message encode/decode and message send/recv is worthwhile, as it makes the API easier to understand.  I might feel a bit differently after I try it with request/response which is used a lot more than the event functions.  What do you think about that one?

**Comment by grondo** [comment]
> I'm not sure what you mean about there being no equivalent flux_event_* using str instead of string?
Sorry, I only meant that you have `flux_msg_get_payload_json_str` for normal messages and `flux_event_decode` for events, both of which return a JSON string, so the naming there seemed slightly inconsistent. I'm sure there was a good reason for it, I just wanted to make a comment so I sounded like I was thinking real hard. :wink:

I think you are right about the clear divisions between encode/decode, send/recv, etc. I think this will get even better with a `flux_msg_t`. My only point was that if *internally* everything uses `json_object *`, then it might be worthwhile to encode some of the laborious boilerplate into an internal-only convenience library that might streamline some code. However, this is easy to do after the fact and isn't high priority as you've already done the work to convert callers. (Again, just trying to add my $.02)

I think your proposal in the comment above sounds great.

**Comment by garlick** [comment]
I just force pushed an update with the following changes:
* drop `_str` suffix from json payload accessors per discussion
* drop `libutil/zdump.[ch]` and add its functionality to `message.[ch]`, and fix handing of message frames containing null terminated strings (they were being displayed in hex)
* drop the `streq` functions from `message.[ch]` as these are no longer needed to avoid copying topic strings on the critical path
* rename `flux_msgtype_string()` to `flux_msg_typestr()` to conform to `message.[ch]` naming.

**Comment by garlick** [comment]
I force pushed couple more changes on friday but forgot to push the button on this comment:

* Drop varargs versions of `flux_event_encode()` (not really that useful)
* Add `flux_msg_fprint()`, `flux_msg_get_route_string()` and fix ping/trace route display to show complete route path, in sender to receiver order.  Drop `libutil/zdump.[ch]`.

**Comment by garlick** [comment]
@trws please review/ACK this PR per discusson in todays project meeting.

The proposed new API is in `libflux/event.h` and `libflux/message.h`.

I think the plan would be to get this merged, then finish excising `json_object` from the public facing API in a follow-on PR.  The second PR would focus on `libflux/request.h` and `libflux/rpc.h`.

File: src/common/libflux/message.h [hunk 1]
```diff
@@ -29,15 +29,18 @@ enum {
  */
 zmsg_t *flux_msg_create (int type);
 
+/* Get/set message type
+ * For FLUX_MSGTYPE_REQUEST: set_type initializes nodeid to FLUX_NODEID_ANY
+ * For FLUX_MSGTYPE_RESPONSE: set_type initializes errnum to 0
+ */
+int flux_msg_set_type (zmsg_t *zmsg, int type);
+int flux_msg_get_type (zmsg_t *zmsg, int *type);
```
Comment (trws) [inline src/common/libflux/message.h]: Should we consider a const qualifier on the zmsg for get calls?

```c
int flux_msg_get_type(const zmsg_t *zmsg, int *type);
```

File: src/common/libflux/event.h [hunk 2]
```diff
@@ -1,22 +1,30 @@
...
+ * If json_str is non-NULL, it is copied to the message payload.
+ * Returns message or NULL on failure with errno set.
+ */
+zmsg_t *flux_event_encode (const char *topic, const char *json_str);
 
 /* Send/receive events
  */
-int flux_event_send (flux_t h, json_object *request, const char *fmt, ...);
-int flux_event_recv (flux_t h, json_object **respp, char **topic, bool nb);
+int flux_event_send (flux_t h, zmsg_t **zmsg);
```
Comment (trws) [inline src/common/libflux/event.h]: The **zmsg makes me curious, is that a side-effect of zmq de-allocating the message on send?

**Comment by trws** [comment]
This looks good to me.  The couple of line comments above are not blockers.

**Comment by garlick** [comment]
@trws: Thanks!

The double-zmsg_t pointer is there because  these functions are ultimately based on `zmsg_send()` which destroys the zmsg on successful send.  I think there is consensus that we don't like this, and that we'll change that in the forthcoming `flux_msg_t` PR.

I'm thinking the const idea should be considered in the same PR.  Probably makes sense.

**Comment by trws** [comment]
Makes sense on both counts, all of that should go in the msg ticket.  The json_object to string changes all look sensible, and seem to have a surprisingly small outward-facing footprint.

**Comment by grondo** [comment]
Thanks for the ACK, Tom!

**Pull request merged**